azurelinuxagent/ga/exthandlers.py

```diff
@@ -1091,28 +1097,61 @@ def get_status_file_path(self, extension=None):
-            ext_status.code = -1
-            ext_status.status = "error"
+            msg = u"Failed to read any status for extension - {0}-{1}, Sequence number {2}. Failed due to {3}"\
+                .format(ext.name, self.ext_handler.properties.version, seq_no, e)
+            ext_status.message = msg
+            ext_status.code = ExtensionErrorCodes.PluginUnknownFailure
+            ext_status.status = EXTENSION_STATUS_WARNING
+
+        if ext_status.code == ExtensionErrorCodes.PluginUnknownFailure:
+            logger.periodic_warn(logger.EVERY_HALF_HOUR, ext_status.message)
+            add_periodic(delta=logger.EVERY_HALF_HOUR, name=ext.name, version=self.ext_handler.properties.version,
+                         op=WALAEventOperation.StatusProcessing, is_success=False, message=ext_status.message,
+                         log_event=False)
+        else:
+            # We did not encounter any PluginUnknownFailures and thus the status file was correctly written and has
+            # valid json.
+            try:
+                parse_ext_status(ext_status, data)
+                if len(data_str) > MAX_STATUS_FILE_SIZE_IN_BYTES:
+                    raise ExtensionStatusError("For Extension Handler {0}-{1}, the status file {2} of size {3} bytes is"
+                                               " too big. Max Limit allowed is {4} bytes"
+                                               .format(ext.name, self.ext_handler.properties.version, ext_status_file,
+                                                       len(data_str), MAX_STATUS_FILE_SIZE_IN_BYTES),
+                                               code=ExtensionStatusError.MaxSizeExceeded)
+            except ExtensionStatusError as e:
+                logger.periodic_warn(logger.EVERY_HALF_HOUR, ustr(e))
+                add_periodic(delta=logger.EVERY_HALF_HOUR, name=ext.name, version=self.ext_handler.properties.version,
+                             op=WALAEventOperation.StatusProcessing, is_success=False, message=ustr(e), log_event=False)
+
+                if e.code == ExtensionStatusError.MaxSizeExceeded:
+                    # Emptying the substatus to reduce the size, and preserve other fields of the text
+                    ext_status.substatusList = []
+                    ext_status.message = ext_status.message if len(ext_status.message) < 200 \
+                        else ext_status.message[:200] + " ... TRUNCATED MESSAGE"
+                elif e.code == ExtensionStatusError.StatusFileMalformed:
+                    ext_status.message = "Could not get a valid status from the extension {0}-{1}. Encountered the " \
+                                         "following error: {0}".format(ext.name, self.ext_handler.properties.version,
+                                                                       ustr(e))
+                    ext_status.code = ExtensionErrorCodes.PluginSettingsStatusInvalid
+                    ext_status.status = EXTENSION_STATUS_WARNING
 
```

Comment:
What happens when `e.code == ExtensionStatusError.StatusFileNotFound`? We would still be returning `ext_status`? Do you need to set the message to explicitly say the status file was not found? How would this be reported to the user?

One other thing; this method is called in `report_ext_status` where there is a check:
````
ext_status = self.collect_ext_status(ext)            
if ext_status is None:
    continue
````

However, this is never true since we will always return `ext_status` which is defined in the beginning of the method as `ExtensionStatus(seq_no=seq_no)`.

Reply:
> However, this is never true since we will always return ext_status which is defined in the beginning of the method as ExtensionStatus(seq_no=seq_no).

This would be true [only] in the case of 
```
seq_no, ext_status_file = self.get_status_file_path(ext)
if seq_no == -1:
    return None
```

Reply:
> What happens when e.code == ExtensionStatusError.StatusFileNotFound? We would still be returning ext_status? Do you need to set the message to explicitly say the status file was not found? How would this be reported to the user?

+1

Reply:
> > This would be true [only] in the case of
> 
> seq_no, ext_status_file = self.get_status_file_path(ext)
> if seq_no == -1:
>     return None

But we still need to report status, right?

Reply:
>But we still need to report status, right?

Yes - and we do send the status.
```
if ext_status.code == ExtensionErrorCodes.PluginUnknownFailure or ext_status.code == \
                ExtensionErrorCodes.PluginSettingsStatusInvalid:
   ... # Log stuff and send telemetry.
else:
   ... # Parse status.

return ext_status
```